### PR TITLE
#17972 and #17975 Fixing PCC and Program Cache issues in Repeat and Expand

### PR DESCRIPTION
**Pull request by jvegaTT**
### Ticket
#17975
#17972

### Problem description
This PR closes two P0 errors by applying bug fixes to the repeat program factory and giving repeat program cache support.

### What's changed
Program factory changes to repeat
Adding Program Cache testing to the CI pipelines for repeat
Removed redundant CI tests in Repeat to help improve CI pipeline times

### Checklist
- [ ] [All post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/all-post-commit-workflows.yaml) CI passes. [Submitted](https://github.com/tenstorrent/tt-metal/actions/workflows/all-post-commit-workflows.yaml) [rerun after PR changes](https://github.com/tenstorrent/tt-metal/actions/runs/13422573676)
- [ ] [Blackhole Post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml) CI passes (if applicable)
- [ ] [Model regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-models.yaml) CI passes (if applicable). [Submitted](https://github.com/tenstorrent/tt-metal/actions/runs/13416833113)
- [ ] [Device performance regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-device-models.yaml) CI passes (if applicable)
- [ ] **(For models and ops writers)** Full [new models tests](https://github.com/tenstorrent/tt-metal/actions/workflows/full-new-models-suite.yaml) CI passes (if applicable)
- [ ] New/Existing tests provide coverage for changes
- [ ] T3K Demo. [Submitted](https://github.com/tenstorrent/tt-metal/actions/runs/13416778070)

### CODE REVIEW

**Review by ntarafdar — APPROVED** [review]

File: ttnn/cpp/ttnn/operations/data_movement/repeat/device/host/repeat_program_factory.cpp [hunk 1]
```diff
@@ -119,7 +119,55 @@ tt::tt_metal::operation::ProgramWithCallbacks rm_repeater_last_dim(
...
+    auto override_runtime_args_callback = [reader_kernel_id](
+                                              const void* operation,
+                                              const tt::tt_metal::Program& program,
+                                              const std::vector<Tensor>& input_tensors,
+                                              const std::vector<std::optional<const Tensor>>&,
+                                              const std::vector<Tensor>& output_tensors) {
+        auto input = input_tensors.at(0);
+        auto output = output_tensors.at(0);
+        ttnn::Shape input_log_shape = ttnn::Shape(input.get_logical_shape().view());
+        ttnn::Shape output_log_shape = ttnn::Shape(output.get_logical_shape().view());
```
Comment (ayerofieiev-tt) [inline ttnn/cpp/ttnn/operations/data_movement/repeat/device/host/repeat_program_factory.cpp:131]: This looks unnecessary complicated

**Review by TT-BrianLiu — CHANGES_REQUESTED** [review]
I don't think your implementation of program callback is correct. You only need to update the args that actually change, but it seems like you copied most of the program setup itself, which is really expensive.

Another thing is, you can capture a lot more args (eg. cores) from the program setup instead of recalculating everything (unless they are expected to change).

File: tests/ttnn/unit_tests/operations/test_repeat.py [hunk 2]
```diff
@@ -12,20 +12,17 @@
 from models.utility_functions import comp_pcc
 from tests.ttnn.utils_for_testing import assert_with_pcc
 
+from tqdm import tqdm
```
[the remark this answers is not in the document]
Comment (jvegaTT) [reply tests/ttnn/unit_tests/operations/test_repeat.py:15]: Yeah in rushing to fix I copy pasted the test code from the issue, I am cleaning it up.

**Review by dmakoviichuk-tt — COMMENTED** [review]
Please add explanation on how does it break repeat calls with the same parameters in a cycle. I currently don't' understand why hash is broken in this case.

File: tests/ttnn/unit_tests/operations/test_repeat.py [hunk 3]
```diff
@@ -75,4 +70,31 @@ def test_repeat(device, layout, dtype, shape, repeat_shape):
...
+        pytest.skip("PyTorch repeat dim must be >= tensor dim (although we can handle this).")
+    device.enable_program_cache()
+    num_iters = 3
+    input_tensors = []
+    torch_results = []
+    for i in range(3):
+        torch_tensor = torch.rand(shape, dtype=torch.bfloat16)
+        torch_results.append(torch_tensor.repeat(repeat_shape))
+        input_tensors.append(ttnn.from_torch(torch_tensor, layout=layout, device=device, dtype=ttnn.bfloat16))
+    for i in range(3):
```
Comment (TT-BrianLiu) [inline tests/ttnn/unit_tests/operations/test_repeat.py:93]: `num_iters` is unused.
`use_program_cache` fixture does same thing as `device.enable_program_cache()`

File: tests/ttnn/unit_tests/operations/test_repeat.py [hunk 4]
```diff
@@ -12,20 +12,17 @@
...
-    (2,),
     (1, 2),
-    (1, 4),
-    (2, 1, 3),
-    (1, 2, 3),
     (4, 3, 2, 1),
     (2, 3, 4, 5, 2),
-    (2, 1, 3, 1, 3, 1),
     (2048,),
 ]
```
[the remark this answers is not in the document]
Comment (aliuTT) [reply tests/ttnn/unit_tests/operations/test_repeat.py:25]: In general, how do you assert that program cache is hit?

[the remark this answers is not in the document]
Comment (jvegaTT) [reply tests/ttnn/unit_tests/operations/test_repeat.py:25]: You have 3 identical tensors in every way except different data is contained called in the exact same way so program cache must be hit

Imperially though the test failed before the fix

[the remark this answers is not in the document]
Comment (TT-BrianLiu) [reply tests/ttnn/unit_tests/operations/test_repeat.py:25]: how is `bfloat8` redundant?

File: ttnn/cpp/ttnn/operations/data_movement/repeat/device/host/repeat_program_factory.cpp [hunk 5]
```diff
@@ -119,7 +119,55 @@ tt::tt_metal::operation::ProgramWithCallbacks rm_repeater_last_dim(
...
+        const uint32_t data_size = input.element_size();
+        tt::tt_metal::IDevice* device = input.device();
+        auto compute_with_storage_grid_size = device->compute_with_storage_grid_size();
+        uint32_t num_cores_x = compute_with_storage_grid_size.x;
+        uint32_t num_cores_y = compute_with_storage_grid_size.y;
+        uint32_t num_cores_total = num_cores_x * num_cores_y;
+        uint32_t read_start_page = 0;
+        tt::tt_metal::Buffer* src_buffer = input.buffer();
+        tt::tt_metal::Buffer* dst_buffer = output.buffer();
+        uint32_t number_of_pages = input_log_shape[-2];
```
Comment (aliuTT) [inline ttnn/cpp/ttnn/operations/data_movement/repeat/device/host/repeat_program_factory.cpp:141]: You're passing in rank 1 tensors as input, does TTNN auto expand to guarantee rank 2?

File: tests/ttnn/unit_tests/operations/test_repeat.py [hunk 6]
```diff
@@ -75,4 +72,27 @@ def test_repeat(device, layout, dtype, shape, repeat_shape):
...
+        torch_tensor = torch.rand(shape, dtype=torch.bfloat16)
+        torch_result = torch_tensor.repeat(repeat_shape)
+        input_tensor = ttnn.from_torch(torch_tensor, layout=layout, device=device, dtype=ttnn.bfloat16)
+        output = ttnn.repeat(input_tensor, ttnn.Shape(repeat_shape))
+        output = ttnn.to_torch(output)
+        assert (
+            output.shape == torch_result.shape
+        ), f"Output shape {output.shape} does not match torch shape {torch_result.shape}"
+
+        assert_with_pcc(torch_result, output, 0.9999)
```
[the remark this answers is not in the document]
Comment (jvegaTT) [reply tests/ttnn/unit_tests/operations/test_repeat.py]: Fixed it, I allocate all the tensors first and then do the processing so they can't overlap

File: tests/ttnn/unit_tests/operations/test_repeat.py [hunk 7]
```diff
@@ -75,4 +70,31 @@ def test_repeat(device, layout, dtype, shape, repeat_shape):
     assert_with_pcc(torch_result, output, 0.9999)
 
 
-# TODO! test program cache when it is implemented
+@pytest.mark.parametrize("layout", layouts)
+@pytest.mark.parametrize("shape", shapes)
+@pytest.mark.parametrize("repeat_shape", repeat_shapes)
+def test_pc_repeat(device, layout, shape, repeat_shape):
```
[the remark this answers is not in the document]
Comment (jvegaTT) [reply tests/ttnn/unit_tests/operations/test_repeat.py:76]: Sure, I tested those tests work but I can directly add them

File: ttnn/cpp/ttnn/operations/data_movement/repeat/device/host/repeat_program_factory.cpp (hunk 5 above)
Comment (jvegaTT) [reply ttnn/cpp/ttnn/operations/data_movement/repeat/device/host/repeat_program_factory.cpp:141]: Yeah, the code in repeat.cpp will expand to rank 2 before viewing it back

File: tests/ttnn/unit_tests/operations/test_repeat.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (jvegaTT) [reply tests/ttnn/unit_tests/operations/test_repeat.py:25]: Ok, adding that one back

File: ttnn/cpp/ttnn/operations/data_movement/repeat/device/host/repeat_program_factory.cpp (hunk 1 above)
Comment (jvegaTT) [reply ttnn/cpp/ttnn/operations/data_movement/repeat/device/host/repeat_program_factory.cpp]: Line has been removed with function simplification

File: ttnn/cpp/ttnn/operations/data_movement/repeat/device/host/repeat_program_factory.cpp [hunk 8]
```diff
@@ -119,7 +119,55 @@ tt::tt_metal::operation::ProgramWithCallbacks rm_repeater_last_dim(
...
+        uint32_t responsibility = ((number_of_pages - 1) / num_cores_total) + 1;
+        uint32_t done = 0;
+        for (int core_x = 0; core_x < num_cores_x; core_x++) {
+            for (int core_y = 0; core_y < num_cores_y; core_y++) {
+                CoreCoord core = {core_x, core_y};
+                if (done == 1) {
```
[the remark this answers is not in the document]
Comment (jvegaTT) [reply ttnn/cpp/ttnn/operations/data_movement/repeat/device/host/repeat_program_factory.cpp]: I have greatly simplified the callback to just recalculate the responsibility and buffers which are the two aspects which can change

File: tests/ttnn/unit_tests/operations/test_repeat.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (TT-BrianLiu) [reply tests/ttnn/unit_tests/operations/test_repeat.py]: Saw the comment about testing program cache. You should assert that the number of generated caches don't change per loop. That's how you guarantee that you are hitting the cache

File: tests/ttnn/unit_tests/operations/test_repeat.py [hunk 9]
```diff
@@ -75,4 +70,81 @@ def test_repeat(device, layout, dtype, shape, repeat_shape):
...
+            output.shape == torch_results[i].shape
+        ), f"Output shape {output.shape} does not match torch shape {torch_results[i].shape}"
+
+        assert_with_pcc(torch_results[i], output, 0.9999)
+
+
+# 17975 test cases
+
+
+def test_17975_a(device, use_program_cache):
```
Comment (ayerofieiev-tt) [inline tests/ttnn/unit_tests/operations/test_repeat.py:105]: not a good test case name

Comment (dmakoviichuk-tt) [reply tests/ttnn/unit_tests/operations/test_repeat.py:105]: something like test_github_17975 is better.
@ayerofieiev-tt  or just comment with link to bug is enought

Comment (ayerofieiev-tt) [reply tests/ttnn/unit_tests/operations/test_repeat.py:105]: same for b, c

Comment (bbradelTT) [reply tests/ttnn/unit_tests/operations/test_repeat.py:105]: The test names need to be descriptive of functionality being tested.
Comments above them can point to specific issues.

**Review by ayerofieiev-tt — APPROVED** [review]

File: tests/ttnn/unit_tests/operations/test_repeat.py (hunk 9 above)
Comment (jvegaTT) [reply tests/ttnn/unit_tests/operations/test_repeat.py:105]: I agree Boris, the problem here is I feel what these tests probe was already covered by previous testing, but another reviewer said I still needed to include the 3 failing cases from the original issue

File: tests/ttnn/unit_tests/operations/test_repeat.py [hunk 10]
```diff
@@ -75,4 +70,81 @@ def test_repeat(device, layout, dtype, shape, repeat_shape):
...
+    y = torch.rand((1, 1, 256, 384), dtype=torch.bfloat16)
+
+    for _ in range(10):
```
Comment (TT-BrianLiu) [inline tests/ttnn/unit_tests/operations/test_repeat.py:108]: 10 seems a bit excessive?

File: tests/ttnn/unit_tests/operations/test_repeat.py [hunk 11]
```diff
@@ -75,4 +70,81 @@ def test_repeat(device, layout, dtype, shape, repeat_shape):
...
+        y_tt = ttnn.from_torch(y, dtype=ttnn.bfloat16, layout=ttnn.TILE_LAYOUT, device=device)
+        print("program cache: ", device.num_program_cache_entries())
```
Comment (TT-BrianLiu) [inline tests/ttnn/unit_tests/operations/test_repeat.py:110]: Remove and add assert on expected program cache entries

File: tests/ttnn/unit_tests/operations/test_repeat.py [hunk 12]
```diff
@@ -75,4 +70,81 @@ def test_repeat(device, layout, dtype, shape, repeat_shape):
...
+
+        x = torch.zeros((64, 1, 256, 384), dtype=torch.bfloat16)
+        x_tt = ttnn.from_torch(x, dtype=ttnn.bfloat16, layout=ttnn.TILE_LAYOUT, device=device)
+
+        z_tt = x_tt + y_tt
+
+        for i in range(64):
```
Comment (TT-BrianLiu) [inline tests/ttnn/unit_tests/operations/test_repeat.py:117]: Please use repeated dim instead of hard coding

File: tests/ttnn/unit_tests/operations/test_repeat.py [hunk 13]
```diff
@@ -75,4 +70,81 @@ def test_repeat(device, layout, dtype, shape, repeat_shape):
...
+            z_torch = ttnn.to_torch(z_tt[i : i + 1])
+            assert torch.allclose(z_torch, y, atol=1e-2), f"z_torch[{i}] != y"
+
+
+def test_17975_b(device, use_program_cache):
```
Comment (TT-BrianLiu) [inline tests/ttnn/unit_tests/operations/test_repeat.py:122]: Same comments as first test. Also can we commonize these tests?

File: tests/ttnn/unit_tests/operations/test_repeat.py [hunk 14]
```diff
@@ -75,4 +70,81 @@ def test_repeat(device, layout, dtype, shape, repeat_shape):
...
+        ttnn.repeat(y_tt, [4, 1, 1, 1])
+        z_tt = ttnn.experimental.add(x_tt, y_tt)
+        # z_tt = x_tt + y_tt
+
+        for i in range(4):
+            z_torch = ttnn.to_torch(z_tt[i : i + 1])
+            assert torch.allclose(z_torch, y, atol=1e-2), f"z_torch[{i}] != y"
+
+
+def test_17975_c(device, use_program_cache):
```
Comment (TT-BrianLiu) [inline tests/ttnn/unit_tests/operations/test_repeat.py:141]: Same as a and b

File: tests/ttnn/unit_tests/operations/test_repeat.py (hunk 12 above)
Comment (jvegaTT) [reply tests/ttnn/unit_tests/operations/test_repeat.py:130]: I am just trying to match the tests from the issue + adding the checks that PC entries is not increasing

**Review by TT-BrianLiu — APPROVED** [review]

File: tests/ttnn/unit_tests/operations/test_repeat.py [hunk 15]
```diff
@@ -75,4 +70,123 @@ def test_repeat(device, layout, dtype, shape, repeat_shape):
...
+            assert (
+                device.num_program_cache_entries() == base_program_cache_entires,
+                "program cache entries differ on same configs",
+            )
+
+
+# 17975 test cases
+
+
+def test_github_issues_from_17975(device, use_program_cache):
```
Comment (bbradelTT) [inline tests/ttnn/unit_tests/operations/test_repeat.py:112]: I still think this test should be renamed with something more descriptive. That can be done later though.

**Review by bbradelTT — APPROVED** [review]

**Review by dmakoviichuk-tt — APPROVED** [review]

**Review by rfurko-tt — COMMENTED** [review]
LGTM, thanks!

**Review by rfurko-tt — APPROVED** [review]
